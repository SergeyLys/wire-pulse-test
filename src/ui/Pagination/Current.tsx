import { memo } from "react";

const Current = ({currentPage, totalPages}: {currentPage?: number, totalPages?: number}) => <div data-testid="pagination-current-page" className="current">{currentPage} of {totalPages}</div>;

export default memo(Current);