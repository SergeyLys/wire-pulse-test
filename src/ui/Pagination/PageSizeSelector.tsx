import { memo } from "react";
import { Dropdown } from "../Dropdown";

type PageSizeOption = {
  id: number;
  value: string;
}

type PageSizeSelectorProps = {
  handleSizeChange?: (pageSize: number) => void;
  pageSizeOptions: PageSizeOption[];
  defaultPageSize?: number;
  styles?: Record<string, any>;
}

const PageSizeSelector: React.FC<PageSizeSelectorProps> = ({handleSizeChange, pageSizeOptions, defaultPageSize, styles}) => {
  return (
    <div style={styles}>
      <Dropdown
        openerTestId="pagination-page-size-selector"
        onChange={(pageSize: number) => handleSizeChange!(pageSize)}
      >
        {pageSizeOptions.map((option: PageSizeOption, i) => (
          <Dropdown.Option testid="pagination-page-size-selector-item" default={option.id === defaultPageSize} key={option.id} value={option}>
            {option.value}
          </Dropdown.Option>
        ))}
      </Dropdown>
    </div>
  )
}

export type {PageSizeOption, PageSizeSelectorProps};
export default memo(PageSizeSelector);