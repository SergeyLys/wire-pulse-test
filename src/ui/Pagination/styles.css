.pagination-container {
  background-color: white;
  padding: 10px 15px;
}

.arrow {
  width: 15px;
  height: 15px;
  cursor: pointer;
}

.arrow svg {
  max-width: 100%;
  height: auto;
}

.current {
  font-size: 18px;
  font-weight: bold;
  margin: 0 5px;
  width: 100px;
  text-align: center;
}