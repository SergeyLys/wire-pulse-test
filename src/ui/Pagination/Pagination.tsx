import React, { memo, useCallback, useEffect, useState } from "react";
import PageSizeSelector, {PageSizeOption} from "./PageSizeSelector";
import Arrow from "./Arrow";
import Current from "./Current";
import './styles.css';

type Props = {
  total: number;
  defaultPageSize: number;
  current: number;
  onChange: (currentPage: number, pageSize: number) => void;
  children?: React.ReactElement | React.ReactElement[];
}

type PaginationComposition = {
  Arrow: typeof Arrow;
  Current: typeof Current;
  PageSizeSelector: typeof PageSizeSelector;
}

const Pagination: React.FC<Props> = ({total, current, defaultPageSize, onChange, children}) => {
  const [pageSize, setPageSize] = useState(defaultPageSize);
  const [currentPage, setCurrentPage] = useState(current);

  const handlePageClick = useCallback((page: number, size: number) => {
    const totalPages = Math.ceil(total / pageSize);

    if (page < 1 || page > totalPages) return;

    setCurrentPage(page);
    setPageSize(size);
    onChange(page, pageSize);
  }, [currentPage, pageSize, total]);

  const handlePageSizeChange = useCallback((size: number) => {
    setPageSize(size);

    if (currentPage * size > total) {
      const nextPage = Math.round(total / size);
      setCurrentPage(nextPage);
      onChange(nextPage, size);
      return;
    }

    onChange(currentPage, size);
  }, [pageSize, total, currentPage]);

  useEffect(() => {
    if (currentPage * pageSize > total) {
      const nextPage = Math.round(total / pageSize) || 1;
      setCurrentPage(nextPage);
    }
  }, [total]);

  return (
    <div className="flex pagination-container align-center">
      {children && React.Children.map(children, (child) => {

        if (child.type === MemoizedPagination.Arrow) {
          return React.cloneElement(child, {
            handlePageClick: () => {
              if (child.props.direction === 'left') {
                handlePageClick(currentPage - 1, pageSize);
              }

              if (child.props.direction === 'right') {
                handlePageClick(currentPage + 1, pageSize);
              }
            },
          });
        }

        if (child.type === MemoizedPagination.Current) {
          return React.cloneElement(child, {
            currentPage,
            totalPages: Math.ceil(total / pageSize)
          });
        }

        if (child.type === MemoizedPagination.PageSizeSelector) {
          return React.cloneElement(child, {
            handleSizeChange: (option: PageSizeOption) => handlePageSizeChange(option.id),
            defaultPageSize
          });
        }

        return null;
      })}
    </div>
  )
}

const MemoizedPagination = memo(Pagination) as React.NamedExoticComponent<Props> & PaginationComposition;

MemoizedPagination.Arrow = Arrow;
MemoizedPagination.Current = Current;
MemoizedPagination.PageSizeSelector = PageSizeSelector;

export type { Props, PageSizeOption }
export default MemoizedPagination;