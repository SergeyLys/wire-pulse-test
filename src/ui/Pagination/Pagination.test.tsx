import renderer from 'react-test-renderer';
import { render, screen, fireEvent } from '@testing-library/react';
import Pagination, { Props, PageSizeOption } from './Pagination';

const Component = (props: Props & {pageSizeOptions: PageSizeOption[]}) => (
  <Pagination {...props}>
    <div data-testid="element-to-trim" />
    <Pagination.PageSizeSelector pageSizeOptions={props.pageSizeOptions} />
    <Pagination.Arrow direction='left' />
    <Pagination.Current />
    <Pagination.Arrow direction='right' />
  </Pagination>
);

const setupComponent = (props?: Record<string, any>) => {
  const defaultProps = {
    total: 100,
    defaultPageSize: 10,
    current: 1,
    onChange: jest.fn(),
    pageSizeOptions: [{id: 10, value: '10'}, {id: 20, value: '20'}, {id: 100, value: '100'}]
  };

  const mergedProps = {...defaultProps, ...props};

  return <Component {...mergedProps} />;
}

describe('Pagination component', () => {
  it('renders correctly', () => {
    const tree = renderer.create(
      setupComponent()
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render only related components', () => {
    render(setupComponent());

    expect(screen.queryByTestId('element-to-trim')).not.toBeInTheDocument();
  });

  it('should increase current page', () => {
    const props = {
      onChange: jest.fn()
    }
    const mockOnChange = jest.spyOn(props, 'onChange');
    render(setupComponent(props));

    fireEvent.click(screen.getByTestId('pagination-arrow-next'));

    expect(mockOnChange).toHaveBeenCalledWith(2, 10);
    expect(screen.getByTestId('pagination-current-page')).toHaveTextContent('2')
  });

  it('should decrease current page', () => {
    const props = {
      current: 10,
      onChange: jest.fn()
    }
    const mockOnChange = jest.spyOn(props, 'onChange');

    render(setupComponent(props));

    fireEvent.click(screen.getByTestId('pagination-arrow-prev'));

    expect(mockOnChange).toHaveBeenCalledWith(9, 10);
    expect(screen.getByTestId('pagination-current-page')).toHaveTextContent('9')
  });

  it('should not decrease current page if it is 1', () => {
    render(setupComponent());

    fireEvent.click(screen.getByTestId('pagination-arrow-prev'));

    expect(screen.getByTestId('pagination-current-page')).toHaveTextContent('1')
  });

  it('should not increase current page if it is last', () => {
    render(setupComponent({current: 10}));

    fireEvent.click(screen.getByTestId('pagination-arrow-next'));

    expect(screen.getByTestId('pagination-current-page')).toHaveTextContent('10')
  });

  it('should render page size correctly', () => {
    render(setupComponent());

    expect(screen.getByTestId('pagination-page-size-selector')).toBeInTheDocument();
  });

  it('should react on page size change', () => {
    const props = {
      onChange: jest.fn()
    }
    const mockOnChange = jest.spyOn(props, 'onChange');
    render(setupComponent(props));

    fireEvent.click(screen.getByTestId('pagination-page-size-selector'));
    fireEvent.click(screen.getAllByTestId('pagination-page-size-selector-item')[1]);
    expect(mockOnChange).toHaveBeenCalledWith(1, 20);
  });

  it('should calculate pages amount if it is higher than total elements', () => {
    const props = {
      onChange: jest.fn(),
      current: 10
    }
    const mockOnChange = jest.spyOn(props, 'onChange');
    render(setupComponent(props));

    fireEvent.click(screen.getByTestId('pagination-page-size-selector'));
    fireEvent.click(screen.getAllByTestId('pagination-page-size-selector-item')[1]);
    expect(mockOnChange).toHaveBeenCalledWith(5, 20);
  });
});
