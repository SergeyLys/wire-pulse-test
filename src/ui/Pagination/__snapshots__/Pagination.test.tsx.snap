// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination component renders correctly 1`] = `
<div
  className="flex pagination-container align-center"
>
  <div>
    <div
      className="dropdown"
    >
      <div
        data-testid="pagination-page-size-selector"
        onClick={[Function]}
      >
        <div
          className="dropdown-button "
        >
          10
        </div>
      </div>
    </div>
  </div>
  <div
    className="prev arrow"
    data-testid="pagination-arrow-prev"
    onClick={[Function]}
  >
    <svg
      height="800px"
      version="1.1"
      viewBox="0 0 34.075 34.075"
      width="800px"
      xmlns="http://www.w3.org/2000/svg"
    >
      <g>
        <g>
          <path
            d="M24.57,34.075c-0.505,0-1.011-0.191-1.396-0.577L8.11,18.432c-0.771-0.771-0.771-2.019,0-2.79 L23.174,0.578c0.771-0.771,2.02-0.771,2.791,0s0.771,2.02,0,2.79l-13.67,13.669l13.67,13.669c0.771,0.771,0.771,2.021,0,2.792 C25.58,33.883,25.075,34.075,24.57,34.075z"
            style={
              Object {
                "fill": "#010002",
              }
            }
          />
        </g>
      </g>
    </svg>
  </div>
  <div
    className="current"
    data-testid="pagination-current-page"
  >
    1
     of 
    10
  </div>
  <div
    className="next arrow"
    data-testid="pagination-arrow-next"
    onClick={[Function]}
  >
    <svg
      height="800px"
      version="1.1"
      viewBox="0 0 185.343 185.343"
      width="800px"
      xmlns="http://www.w3.org/2000/svg"
    >
      <g>
        <g>
          <path
            d="M51.707,185.343c-2.741,0-5.493-1.044-7.593-3.149c-4.194-4.194-4.194-10.981,0-15.175 l74.352-74.347L44.114,18.32c-4.194-4.194-4.194-10.987,0-15.175c4.194-4.194,10.987-4.194,15.18,0l81.934,81.934 c4.194,4.194,4.194,10.987,0,15.175l-81.934,81.939C57.201,184.293,54.454,185.343,51.707,185.343z"
            style={
              Object {
                "fill": "#010002",
              }
            }
          />
        </g>
      </g>
    </svg>
  </div>
</div>
`;
